protocol/consistencychecks.go

```diff
@@ -0,0 +1,379 @@
+// and SignedTreeRoot value, as well as directory's policies
+// (e.g., wherether the TemporaryBinding extension is being used or not)
+// The client should create a new ConsistencyChecks instance only once,
+// when she registers her binding with a ConiksDirectory.
+// This ConsistencyChecks instance then will be used to verify
+// the returned responses from the ConiksDirectory.
+type ConsistencyChecks struct {
+	SavedSTR *m.SignedTreeRoot
+
+	// extensions settings
+	useTBs bool
+	TBs    map[string]*m.TemporaryBinding
+
+	// signing key
+	signKey sign.PublicKey
+}
+
+// NewCC creates an instance of ConsistencyChecks using
+// the pinning directory's STR at epoch 0.
+func NewCC(savedSTR *m.SignedTreeRoot, useTBs bool, signKey sign.PublicKey) *ConsistencyChecks {
+
+	// TODO: see #110
+	if !useTBs {
+		panic("Currently the server is forced to use TBs")
+	}
+
+	cc := &ConsistencyChecks{
+		SavedSTR: savedSTR,
+		useTBs:   useTBs,
+		signKey:  signKey,
+	}
+
+	if useTBs {
+		cc.TBs = make(map[string]*m.TemporaryBinding)
+	}
+	return cc
+}
+
+// clone makes a copy of the current consistency state.
+func (cc *ConsistencyChecks) clone() *ConsistencyChecks {
+	newCC := &ConsistencyChecks{
+		SavedSTR: &(*cc.SavedSTR),
+		useTBs:   cc.useTBs,
+		signKey:  cc.signKey,
+	}
+
+	if cc.useTBs {
+		newCC.TBs = make(map[string]*m.TemporaryBinding)
+		for k, v := range cc.TBs {
+			newCC.TBs[k] = v
+		}
+	}
+
+	return newCC
+}
+
+func (cc *ConsistencyChecks) update(updated *ConsistencyChecks) {
+	cc.SavedSTR = updated.SavedSTR
+	cc.TBs = updated.TBs
+}
+
+// Verify verifies the directory's response for a request and
+// updates the current state if all the verifications are successful.
+// It makes a copy of the current state and does verifications
+// using the cloned state. To verify the directory's response,
+// it first verifies the directory's returned status code of the request.
+// If the status code is not in ErrorResponses array, it means
+// the directory has successfully handled the request.
+// The verifier will then verify the consistency state of the response.
+// During the verifications, the cloned state would be modified.
+// This cloned state will be used to update the original
+// state if all the verifications are passed, otherwise it will be discarded.
+func (cc *ConsistencyChecks) Verify(requestType int, msg *Response,
+	uname string, key []byte) ErrorCode {
+	if ErrorResponses[msg.Error] {
+		return msg.Error
+	}
+
+	var verifResult error
+	cloneCC := cc.clone()
+
+	switch requestType {
+	case RegistrationType:
+		verifResult = cloneCC.verifyRegistration(requestType, msg, uname, key)
+	case KeyLookupType:
+		verifResult = cloneCC.verifyKeyLookup(requestType, msg, uname, key)
+	case MonitoringType:
+	case KeyLookupInEpochType:
+	default:
+		panic("[coniks] Unknown request type")
+	}
+
+	if verifResult == PassedWithAProofOfAbsence ||
+		verifResult == PassedWithAProofOfInclusion {
+		cc.update(cloneCC)
```

Comment:
I agree. I'm still not sure we even need to be operating on a cloned version of the consistency checks state since I'm still unclear how a client will use the information in the state.